In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
# se utiliza para el manejo de rutas y directorios.
import os

# Calculo cientifico y vectorial para python
import numpy as np

# Librerias para graficar
from matplotlib import pyplot

# Modulo de optimización de scipy
from scipy import optimize

# le dice a matplotlib que incruste gráficos en el cuaderno
%matplotlib inline

import pandas as pd
from sklearn.preprocessing import LabelEncoder
from datetime import datetime

# Importar libreria para dividir entrenamiento y prueba
from sklearn.model_selection import train_test_split

In [3]:
data = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Datasets/TMNIST_Data.csv')

In [4]:
print(data)

                                  names  labels  1  2  3  4  5  6  7  8  ...  \
0                    GrandHotel-Regular       2  0  0  0  0  0  0  0  0  ...   
1            EncodeSansCondensed-Medium       8  0  0  0  0  0  0  0  0  ...   
2                        Varela-Regular       4  0  0  0  0  0  0  0  0  ...   
3                        ArefRuqaa-Bold       3  0  0  0  0  0  0  0  0  ...   
4                     KulimPark-Regular       1  0  0  0  0  0  0  0  0  ...   
...                                 ...     ... .. .. .. .. .. .. .. ..  ...   
29895                     Raleway[wght]       7  0  0  0  0  0  0  0  0  ...   
29896                     Charm-Regular       5  0  0  0  0  0  0  0  0  ...   
29897  BigShouldersStencilDisplay-Black       0  0  0  0  0  0  0  0  0  ...   
29898                  Gabriela-Regular       1  0  0  0  0  0  0  0  0  ...   
29899                       Ovo-Regular       6  0  0  0  0  0  0  0  0  ...   

       775  776  777  778  779  780  78

Conversion de la columna que esta en objeto a numeros entreros

In [5]:
#variables categoricas
columnas_categoricas = data.select_dtypes(include=['object']).columns

In [6]:
#Procesamiento de frases o palabras
for columna in columnas_categoricas:
  le = LabelEncoder()
  data[columna] = le.fit_transform(data[columna])

In [7]:
print(data)

       names  labels  1  2  3  4  5  6  7  8  ...  775  776  777  778  779  \
0       1097       2  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
1        851       8  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
2       2908       4  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
3        156       3  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
4       1539       1  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
...      ...     ... .. .. .. .. .. .. .. ..  ...  ...  ...  ...  ...  ...   
29895   2368       7  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
29896    624       5  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
29897    463       0  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
29898   1030       1  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
29899   2166       6  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   

       780  781  782  783  784  
0        0    0    0    0    0

Para la division del dataset en entrenamiento y prueba del 80% y 20% respectivamente se aplico la libreria train_test_split que con eso nos posibilita crear los nuevos datasets

In [8]:
# Division de dataset en entrenamiento y prueba
entrenamiento_data, prueba_data = train_test_split(data, test_size=0.2, random_state=42)

In [9]:
print(entrenamiento_data)

       names  labels  1  2  3  4  5  6  7  8  ...  775  776  777  778  779  \
1256     257       9  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
15610    494       0  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
4716     676       8  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
12927   1358       5  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
10980   2058       8  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
...      ...     ... .. .. .. .. .. .. .. ..  ...  ...  ...  ...  ...  ...   
29802    314       1  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
5390    2873       6  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
860     2848       1  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
15795    534       7  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
23654   1105       1  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   

       780  781  782  783  784  
1256     0    0    0    0    0

In [10]:
print(prueba_data)

       names  labels  1  2  3  4  5  6  7  8  ...  775  776  777  778  779  \
15756   1951       3  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
19617   1224       5  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
18973   2030       8  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
16397    581       9  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
16366   2830       0  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
...      ...     ... .. .. .. .. .. .. .. ..  ...  ...  ...  ...  ...  ...   
708     1738       9  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
23587   2393       8  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
27362    739       3  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
20592   1791       8  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   
9934    1898       5  0  0  0  0  0  0  0  0  ...    0    0    0    0    0   

       780  781  782  783  784  
15756    0    0    0    0    0

Creacion de los nuevos datasets entrenamiento y prueba


In [11]:
entrenamiento_data.to_csv("entrenamiento.csv", index=False)

In [12]:
prueba_data.to_csv("prueba.csv", index=False)

Carga del datasets del entrenamiento detectando que nuestra variable y que son nuestras etiquetas se encuentra en la segunda columna

In [13]:
data = np.loadtxt("/content/entrenamiento.csv", delimiter=',', skiprows=1)
#print(data)
#X, y = data[:,2:], data[:, 1]
#print(X.shape)
#print(y.shape)

In [14]:
# En caso de identificar la "y" en otra posicion, no ultima
columnas_seleccionadas = list(range(0, 1)) + list(range(2, 786))
X = data[:, columnas_seleccionadas]
y = data[:, 1]
m = y.size

In [15]:
print(X.shape)
print(y.shape)

(23920, 785)
(23920,)


In [16]:
y[y == 10] = 0

In [17]:
input_layer_size = 786
num_labels = 10
m = y.size

In [18]:
# Selecciona aleatoriamente 100 puntos de datos para mostrar
rand_indices = np.random.choice(m, 100, replace=False)
sel = X[rand_indices, :]

#displayData(sel)

In [19]:
def sigmoid(z):
    """
    Calcula la sigmoide de z.
    """
    return 1.0 / (1.0 + np.exp(-z))

In [20]:
def lrCostFunction(theta, X, y, lambda_):
    """
    Calcula el costo de usar theta como parámetro para la regresión logística regularizada y
    el gradiente del costo w.r.t. a los parámetros.

    Parametros
    ----------
    theta : array_like
        Parametro theta de la regresion logistica. Vector de la forma(shape) (n, ). n es el numero de caracteristicas
        incluida la intercepcion

    X : array_like
        Dataset con la forma(shape) (m x n). m es el numero de ejemplos, y n es el numero de
        caracteristicas (incluida la intercepcion).

    y : array_like
        El conjunto de etiquetas. Un vector con la forma (shape) (m, ). m es el numero de ejemplos

    lambda_ : float
        Parametro de regularización.

    Devuelve
    -------
    J : float
        El valor calculado para la funcion de costo regularizada.

    grad : array_like
        Un vector de la forma (shape) (n, ) que es el gradiente de la
        función de costo con respecto a theta, en los valores actuales de theta..
    """
#     alpha = 0.003
#     theta = theta.copy()
    # Inicializa algunos valores utiles
    m = y.size

    # convierte las etiquetas a valores enteros si son boleanos
    if y.dtype == bool:
        y = y.astype(int)

    J = 0
    grad = np.zeros(theta.shape)

    h = sigmoid(X.dot(theta.T))

    temp = theta
    temp[0] = 0

#     J = (1 / m) * np.sum(-y.dot(np.log(h)) - (1 - y).dot(np.log(1 - h)))
    J = (1 / m) * np.sum(-y.dot(np.log(h)) - (1 - y).dot(np.log(1 - h))) + (lambda_ / (2 * m)) * np.sum(np.square(temp))

    grad = (1 / m) * (h - y).dot(X)
#     theta = theta - (alpha / m) * (h - y).dot(X)
    grad = grad + (lambda_ / m) * temp

    return J, grad
#    return J, theta

In [21]:
def oneVsAll(X, y, num_labels, lambda_):
    """
    Trains num_labels logistic regression classifiers and returns
    each of these classifiers in a matrix all_theta, where the i-th
    row of all_theta corresponds to the classifier for label i.

    Parameters
    ----------
    X : array_like
        The input dataset of shape (m x n). m is the number of
        data points, and n is the number of features. Note that we
        do not assume that the intercept term (or bias) is in X, however
        we provide the code below to add the bias term to X.

    y : array_like
        The data labels. A vector of shape (m, ).

    num_labels : int
        Number of possible labels.

    lambda_ : float
        The logistic regularization parameter.

    Returns
    -------
    all_theta : array_like
        The trained parameters for logistic regression for each class.
        This is a matrix of shape (K x n+1) where K is number of classes
        (ie. `numlabels`) and n is number of features without the bias.
    """
    # algunas variables utiles
    m, n = X.shape

    all_theta = np.zeros((num_labels, n + 1))

    # Agrega unos a la matriz X
    X = np.concatenate([np.ones((m, 1)), X], axis=1)

    for c in np.arange(num_labels):
        initial_theta = np.zeros(n + 1)
        options = {'maxiter': 50}
        res = optimize.minimize(lrCostFunction,
                                initial_theta,
                                (X, (y == c), lambda_),
                                jac=True,
                                method='CG',
                                options=options)

        all_theta[c] = res.x

    return all_theta

Carga de nuestro datasets prueba para iniciar nuestras predicciones


In [22]:
data = np.loadtxt("/content/prueba.csv", delimiter=',', skiprows=1)

In [23]:
# En caso de identificar la "y" en otra posicion, no ultima
columnas_seleccionadas = list(range(0, 1)) + list(range(2, 786))
X = data[:, columnas_seleccionadas]
y = data[:, 1]
m = y.size

In [24]:
print(X.shape)
print(y.shape)

(5980, 785)
(5980,)


In [25]:
lambda_ = 0.001
all_theta = oneVsAll(X, y, num_labels, lambda_)
print(all_theta.shape)

<ipython-input-19-575a438e2fc3>:5: RuntimeWarning: overflow encountered in exp
  return 1.0 / (1.0 + np.exp(-z))
<ipython-input-20-cf6e0809223e>:49: RuntimeWarning: divide by zero encountered in log
  J = (1 / m) * np.sum(-y.dot(np.log(h)) - (1 - y).dot(np.log(1 - h))) + (lambda_ / (2 * m)) * np.sum(np.square(temp))
<ipython-input-20-cf6e0809223e>:49: RuntimeWarning: divide by zero encountered in log
  J = (1 / m) * np.sum(-y.dot(np.log(h)) - (1 - y).dot(np.log(1 - h))) + (lambda_ / (2 * m)) * np.sum(np.square(temp))
<ipython-input-20-cf6e0809223e>:49: RuntimeWarning: divide by zero encountered in log
  J = (1 / m) * np.sum(-y.dot(np.log(h)) - (1 - y).dot(np.log(1 - h))) + (lambda_ / (2 * m)) * np.sum(np.square(temp))
<ipython-input-19-575a438e2fc3>:5: RuntimeWarning: overflow encountered in exp
  return 1.0 / (1.0 + np.exp(-z))
<ipython-input-20-cf6e0809223e>:49: RuntimeWarning: divide by zero encountered in log
  J = (1 / m) * np.sum(-y.dot(np.log(h)) - (1 - y).dot(np.log(1 - h))) + 

(10, 786)


In [26]:
def predictOneVsAll(all_theta, X):
    """
    Devuelve un vector de predicciones para cada ejemplo en la matriz X.
    Tenga en cuenta que X contiene los ejemplos en filas.
    all_theta es una matriz donde la i-ésima fila es un vector theta de regresión logística entrenada para la i-ésima clase.
    Debe establecer p en un vector de valores de 0..K-1 (por ejemplo, p = [0, 2, 0, 1]
    predice clases 0, 2, 0, 1 para 4 ejemplos).

    Parametros
    ----------
    all_theta : array_like
        The trained parameters for logistic regression for each class.
        This is a matrix of shape (K x n+1) where K is number of classes
        and n is number of features without the bias.

    X : array_like
        Data points to predict their labels. This is a matrix of shape
        (m x n) where m is number of data points to predict, and n is number
        of features without the bias term. Note we add the bias term for X in
        this function.

    Devuelve
    -------
    p : array_like
        The predictions for each data point in X. This is a vector of shape (m, ).
    """

    m = X.shape[0];
    num_labels = all_theta.shape[0]

    p = np.zeros(m)

    # Add ones to the X data matrix
    X = np.concatenate([np.ones((m, 1)), X], axis=1)
    p = np.argmax(sigmoid(X.dot(all_theta.T)), axis = 1)

    return p

In [27]:
print(X.shape)
pred = predictOneVsAll(all_theta, X)
print('Precision del conjuto de entrenamiento: {:.2f}%'.format(np.mean(pred == y) * 100))
XPrueba = X[10:150, :].copy()
print(XPrueba.shape)
#print(np.ones((1)))
#print(XPrueba)
#p = np.zeros(1)
XPrueba = np.concatenate([np.ones((140, 1)), XPrueba], axis=1)
print(XPrueba.shape)
p = np.argmax(sigmoid(XPrueba.dot(all_theta.T)), axis = 1)
print(p)

#displayData(X[1002:1003, :])
print(y[10:150])

(5980, 785)
Precision del conjuto de entrenamiento: 98.70%
(140, 785)
(140, 786)
[1 7 2 7 0 8 3 0 0 9 9 1 5 5 0 5 5 6 3 5 9 8 5 4 8 3 7 3 5 9 8 4 0 9 7 6 4
 9 8 2 1 5 5 6 6 3 4 2 4 3 2 2 1 8 3 8 9 1 7 2 7 7 1 7 7 4 4 0 0 8 7 1 4 1
 5 1 4 8 0 9 6 7 7 4 0 0 5 5 5 0 8 2 2 4 2 3 4 4 5 2 5 0 3 4 0 2 0 3 8 0 6
 4 0 3 1 3 2 5 7 9 6 2 7 3 3 5 6 4 8 8 0 1 0 9 0 4 5 3 4 7]
[1. 7. 2. 7. 0. 8. 3. 0. 0. 9. 9. 1. 5. 5. 0. 5. 5. 6. 3. 5. 9. 8. 5. 4.
 8. 3. 7. 3. 5. 9. 8. 4. 0. 9. 7. 6. 4. 9. 8. 2. 1. 5. 5. 6. 6. 3. 4. 2.
 4. 3. 2. 2. 1. 7. 3. 8. 0. 1. 7. 2. 7. 7. 1. 7. 7. 4. 4. 0. 0. 8. 7. 2.
 4. 1. 5. 1. 4. 8. 0. 9. 6. 7. 7. 4. 0. 0. 5. 5. 5. 0. 8. 2. 2. 4. 2. 3.
 4. 4. 5. 2. 5. 0. 3. 4. 0. 2. 0. 3. 8. 0. 6. 4. 0. 3. 1. 9. 2. 5. 7. 9.
 6. 2. 7. 3. 3. 5. 6. 4. 8. 8. 0. 1. 0. 9. 0. 4. 5. 3. 4. 7.]


Para la precision del modelo pudimos obtener un 98.70 por ciento que es muy aceptable y demuestra que nuestras predicciones son correctas